File: Assets/ToLua/Lua/user/ui/player/view/change_profession_view.lua
local t = {}
local PREFAB_PATH = 'ui/change_profession/change_profession_view'

local global_data = gamemanager.GetData('global_data')
local game_model = gamemanager.GetModel('game_model')
local role_model = gamemanager.GetModel('role_model')
local player_model = gamemanager.GetModel('player_model')
local player_controller = gamemanager.GetCtrl('player_controller')
local common_skill_icon = require('ui/common_icon/common_skill_icon')
local player_info = require('ui/player/model/player_info')
local hero_data = gamemanager.GetModel('hero_data')

local profession_item = require('ui/player/view/profession_item')

t.isSkillsPanelShowing = true
t.professionItems = {}

function t.Open ()
  local gameObject = Logic.UI.UIMgr.instance:Open(PREFAB_PATH,EUISortingLayer.MainUI, UIOpenMode.Replace)
  t.transform = gameObject:GetComponent(typeof(Transform)) 
   
  local common_top_bar = require ('ui/common_top_bar/common_top_bar')
  common_top_bar = common_top_bar:Create(t.transform:Find('core'))
  common_top_bar.transform:SetAsFirstSibling()
  common_top_bar:SetAsCommonStyle(Common.Localization.LocalizationController.instance:Get('ui.change_profession_view.change_profession_title'), t.ClickCloseButton, true, true, true, true, false, false, false)
  
  t.InitComponent()
  t.BindDelegate()
end

function t.InitComponent ()
  t.roleModelRootTransform = t.transform:Find('core/player_model_root/model_root')
  t.activateProfessionFXGameObject = t.transform:Find('core/player_model_root/ui_shimojihuo').gameObject
  t.activateProfessionFXGameObject:SetActive(false)
  
  t.changeProfessionText = t.transform:Find('core/btn_change_profession/text'):GetComponent(typeof(Text))
  t.changeProfessionButton = t.transform:Find('core/btn_change_profession'):GetComponent(typeof(Button))
  t.changeProfessionButton.onClick:AddListener(t.ClickChangeProfessionButton)
  
  t.infoFrameTransform = t.transform:Find('core/right_anchor/img_info_frame')
  
  t.professionNameText = t.infoFrameTransform:Find('text_profession_name'):GetComponent(typeof(Text))
  t.playerLevelText = t.infoFrameTransform:Find('text_player_level'):GetComponent(typeof(Text))
  t.playerLevelMaxText = t.infoFrameTransform:Find('text_player_level_max'):GetComponent(typeof(Text))
  t.playerExpSlider = t.infoFrameTransform:Find('slider_player_exp'):GetComponent(typeof(Slider))
  t.playerExpPercentText = t.infoFrameTransform:Find('text_player_exp_percent'):GetComponent(typeof(Text))
  
  t.showHideSkillButton = t.infoFrameTransform:Find('skills_root/skills_panel_root/btn_bar'):GetComponent(typeof(Button))
  t.showHideSkillButton.onClick:AddListener(t.ClickShowHideSkillsPanel)
  t.skillsPanelShowHideArrowTransform = t.showHideSkillButton.transform:Find('img_arrow')
  t.skillsPanelRootGameObject = t.infoFrameTransform:Find('skills_root/skills_panel_root').gameObject
  t.skillPanelHidePosTransform = t.infoFrameTransform:Find('skills_root/skills_panel_hide_pos')
  t.skillPanelShowPosTransform = t.infoFrameTransform:Find('skills_root/skills_panel_show_pos')

  t.hpText = t.infoFrameTransform:Find('basic_info_root/text_hp'):GetComponent(typeof(Text))
  t.attackText = t.infoFrameTransform:Find('basic_info_root/text_attack'):GetComponent(typeof(Text))
  t.defenceText = t.infoFrameTransform:Find('basic_info_root/text_defence'):GetComponent(typeof(Text))
  t.speedText = t.infoFrameTransform:Find('basic_info_root/text_speed'):GetComponent(typeof(Text))
  t.critText = t.infoFrameTransform:Find('basic_info_root/text_critical'):GetComponent(typeof(Text))
  t.dodgeText = t.infoFrameTransform:Find('basic_info_root/text_dodge'):GetComponent(typeof(Text))
  t.blockText = t.infoFrameTransform:Find('basic_info_root/text_block'):GetComponent(typeof(Text))
  t.hitText = t.infoFrameTransform:Find('basic_info_root/text_hit'):GetComponent(typeof(Text))
  
  -- [[ Skill Buttons ]] --
  local skillIconsRoot = t.infoFrameTransform:Find('skills_root/skills_panel_root/skills_root')
  t.skillIconTable  = {}
  t.skillIconTable[1] = common_skill_icon.BindTransform(skillIconsRoot:Find('img_skill_01'))
  t.skillIconTable[2] = common_skill_icon.BindTransform(skillIconsRoot:Find('img_skill_02'))
  t.skillIconTable[3] = common_skill_icon.BindTransform(skillIconsRoot:Find('passive/img_skill_03'))
  -- [[ Skill Buttons ]] --
  
  t.GenerateProfessionButtons()
  t.Refresh()
end

function t.BindDelegate ()
  player_model.OnPlayerInfoUpdateDelegate:AddListener(t.OnPlayerInfoUpdateHandler)
  player_model.OnPlayerActivateProfessionSuccessDelegate:AddListener(t.OnPlayerActivateProfessionSuccessHandler)
end

function t.UnbindDelegate ()
  player_model.OnPlayerInfoUpdateDelegate:RemoveListener(t.OnPlayerInfoUpdateHandler)
  player_model.OnPlayerActivateProfessionSuccessDelegate:RemoveListener(t.OnPlayerActivateProfessionSuccessHandler)
end

function t.DespawnCharacter()
  if t.characterEntity then
    PoolController.instance:Despawn(t.characterEntity.name,t.characterEntity)
    t.characterEntity = nil
  end
end

function t.GenerateProfessionButtons ()
  t.professionButtonsRoot = t.infoFrameTransform:Find('professions_root/scrollrect/profession_buttons_root')
  t.professionButtonPrefab = t.infoFrameTransform:Find('professions_root/scrollrect/profession_buttons_root/btn_profession')
  t.professionButtonPrefab.gameObject:SetActive(false)
  
  local currentPlayerInfo = game_model.playerInfo
  t.professionItems = {}
  
  local currentPlayerProfessionButton = Instantiate(t.professionButtonPrefab)
  local currentPlayerProfessionItem = profession_item.NewByGameObject(currentPlayerProfessionButton, currentPlayerInfo)
  currentPlayerProfessionItem.transform:GetComponent(typeof(Button)).onClick:AddListener(function ()
          t.ClickProfessionItem (currentPlayerProfessionItem)
        end)
  table.insert(t.professionItems, currentPlayerProfessionItem)
  currentPlayerProfessionButton.transform:SetParent(t.professionButtonsRoot, false)
  currentPlayerProfessionButton.gameObject:SetActive(true)
  t.ClickProfessionItem (currentPlayerProfessionItem)
  
  for k, v in pairs(currentPlayerInfo.playerData:GetChangeProfessionPlayerDataList ()) do
    --local playerInfo = player_model.GetPlayerDataCorrespondingPlayerInfo(v.id)
  --  if playerInfo == nil then
     local playerInfo = player_info:New(0,  v.id, currentPlayerInfo.hairCutIndex,  currentPlayerInfo.hairColorIndex, currentPlayerInfo.faceIndex, currentPlayerInfo.name, currentPlayerInfo.skinIndex)
     playerInfo.exp = currentPlayerInfo.exp
     playerInfo.level = currentPlayerInfo.level
     playerInfo.strengthenLevel = currentPlayerInfo.strengthenLevel
     playerInfo.strengthenExp = currentPlayerInfo.strengthenExp
     playerInfo.breakthroughLevel = currentPlayerInfo.breakthroughLevel
    --end
    local professionButton = Instantiate(t.professionButtonPrefab)
    local professionItem = profession_item.NewByGameObject(professionButton, playerInfo)
    professionItem.transform:GetComponent(typeof(Button)).onClick:AddListener(function ()
          t.ClickProfessionItem (professionItem)
        end)
    table.insert(t.professionItems, professionItem)
    professionButton.transform:SetParent(t.professionButtonsRoot, false)
    professionButton.gameObject:SetActive(true)
  end
end

function t.RefreshModel()
  LeanTween.delayedCall(0.05, Action(t.RefreshModelDelay)) 
end

function t.RefreshModelDelay()
  t.DespawnCharacter()
  ui_util.ClearChildren(t.roleModelRootTransform,true)
  t.characterEntity = CharacterEntity.CreatePlayerEntityAsUIElementByPlayerInfoLuaTable(t.selectedPlayerInfo, t.roleModelRootTransform, true, true)
  AnimatorUtil.CrossFade(t.characterEntity.anim,AnimatorUtil.VICOTRY_ID,0.3)
end

function t.RefreshChangeProfessionButton ()
  if player_model.IsPlayerUnlocked(t.selectedPlayerInfo.playerData.id) then
    t.changeProfessionText.text = Common.Localization.LocalizationController.instance:Get('ui.change_profession_view.change_profession')
  else
    t.changeProfessionText.text = Common.Localization.LocalizationController.instance:Get('ui.change_profession_view.transfer_profession')
  end
  
  local shouldShowchangeProfessionButton = not (t.selectedPlayerInfo.playerData.id == game_model.playerInfo.playerData.id)
  t.changeProfessionButton.gameObject:SetActive(shouldShowchangeProfessionButton)
end

function t.RefreshPlayerInfos ()
  t.professionNameText.text = t.selectedPlayerInfo.heroData:GetNameWithQualityColor()
  t.playerLevelText.text = t.selectedPlayerInfo.level
  t.playerLevelMaxText.text = t.selectedPlayerInfo:MaxLevel()
  t.playerExpSlider.value = t.selectedPlayerInfo:ExpPercent()
  t.playerExpPercentText.text = string.format('%d%%',math.floor(t.selectedPlayerInfo:ExpPercent()*100))

  -- [[ Player Attr ]] --
  local attrList = role_model.CalcRoleAttributesDic(t.selectedPlayerInfo)
  local equipAddAttrList = role_model.CalcRoleAttributesDicByEquip(t.selectedPlayerInfo)
  local showAttrList = {}
  showAttrList[1] = attrList[RoleAttributeType.HP]
  if game_model.playerInfo:GetRoleAttackAttributeType() == RoleAttackAttributeType.PhysicalAttack then
    showAttrList[2] = attrList[RoleAttributeType.NormalAtk]
  else
    showAttrList[2] = attrList[RoleAttributeType.MagicAtk]
  end
  showAttrList[3] = attrList[RoleAttributeType.Def]
  showAttrList[4] = attrList[RoleAttributeType.Speed]
  showAttrList[5] = attrList[RoleAttributeType.Crit]
  showAttrList[6] = attrList[RoleAttributeType.Dodge]
  showAttrList[7] = attrList[RoleAttributeType.Block]
  showAttrList[8] = attrList[RoleAttributeType.Hit]
  
  t.hpText.text = showAttrList[1]:GetValueString()
  t.attackText.text = showAttrList[2]:GetValueString()
  t.defenceText.text = showAttrList[3]:GetValueString()
  t.speedText.text = showAttrList[4]:GetValueString()
  t.critText.text = showAttrList[5]:GetValueString()
  t.dodgeText.text = showAttrList[6]:GetValueString()
  t.blockText.text = showAttrList[7]:GetValueString()
  t.hitText.text = showAttrList[8]:GetValueString()
  -- [[ Player Attr ]] --
end

function t.RefreshSkill ()
  if t.selectedPlayerInfo == nil then
    return
  end
  
  t.skillIconTable[1]:SetSkillData(t.selectedPlayerInfo.heroData.skillId1, t.selectedPlayerInfo.advanceLevel, t.selectedPlayerInfo.heroData.starMin,false)
  t.skillIconTable[2]:SetSkillData(t.selectedPlayerInfo.heroData.skillId2, t.selectedPlayerInfo.advanceLevel, t.selectedPlayerInfo.heroData.starMin,false)
  t.skillIconTable[3]:SetSkillData(t.selectedPlayerInfo.heroData.passiveId1, t.selectedPlayerInfo.advanceLevel, t.selectedPlayerInfo.heroData.starMin,true)
end

function t.RefreshProfessionItems ()
  for k, v in pairs(t.professionItems) do
    v:Refresh()
    if v.playerInfo.instanceID == game_model.playerInfo.instanceID then
      v.transform:SetAsFirstSibling()
    end
  end
end

function t.Refresh ()
  t.RefreshModel()
  t.RefreshChangeProfessionButton()
  t.RefreshPlayerInfos()
  t.RefreshSkill()
  t.RefreshProfessionItems ()
end

function t.Close ()
  Logic.UI.UIMgr.instance:Close(PREFAB_PATH)
  t.DespawnCharacter()
  t.UnbindDelegate()
end

-- [[ proxy callback ]] --
function t.OnPlayerInfoUpdateHandler ()
  t.Refresh()
end

function t.OnPlayerActivateProfessionSuccessHandler ()
  t.activateProfessionFXGameObject:SetActive(false)
  t.activateProfessionFXGameObject:SetActive(true)
end
-- [[ proxy callback ]] --

-- [[ UI event handles ]] --
function t.ClickCloseButton ()
  t.Close()
end

function t.ClickShowHideSkillsPanel ()
  if t.isSkillsPanelShowing then
    LeanTween.moveLocalY(t.skillsPanelRootGameObject, t.skillPanelHidePosTransform.localPosition.y, 0.2)
    t.skillsPanelShowHideArrowTransform.localRotation = Quaternion.Euler(0, 0, 180)
  else
    LeanTween.moveLocalY(t.skillsPanelRootGameObject, t.skillPanelShowPosTransform.localPosition.y, 0.2)
    t.skillsPanelShowHideArrowTransform.localRotation = Quaternion.Euler(0, 0, 0)
  end
  t.isSkillsPanelShowing = not t.isSkillsPanelShowing
end

function t.ClickProfessionItem (professionItem)
  t.selectedPlayerInfo = professionItem.playerInfo
  for k, v in pairs(t.professionItems) do
      v:SetSelected(false)
  end

  professionItem:SetSelected(true)
  t.Refresh()
end

function t.ClickChangeProfessionButton ()
  if player_model.IsPlayerUnlocked(t.selectedPlayerInfo.playerData.id) then
    local playerDataId = t.selectedPlayerInfo.playerData.id
    player_controller.PlayerChangeReq (player_model.GetPlayerDataCorrespondingPlayerInfo(playerDataId).instanceID)
  else
    LuaInterface.LuaCsTransfer.OpenActivateProfessionView(t.selectedPlayerInfo.playerData.id)
  end
end

function t.ClickChangeProfessionStrengthenButton ()
end
-- [[ UI event handles ]] --

return t

File: Assets/Code/Logic/UI/ChangeProfession/View/ChangeProfessionView.cs
using UnityEngine;
using UnityEngine.UI;
using System.Collections.Generic;
using Common.Util;
using Common.Localization;
using Logic.Character;
using Logic.Game.Model;
using Logic.Player.Model;
using Logic.Player.Controller;
using Logic.Player;
using Logic.Enums;
using Logic.Hero.Model;
using Common.ResMgr;
using Logic.Skill.Model;
using Logic.UI.CommonTopBar.View;
using Logic.UI.Description.View;

namespace Logic.UI.ChangeProfession.View
{
    public class ChangeProfessionView : MonoBehaviour
    {
        public const string PREFAB_PATH = "ui/change_profession/change_profession_view";

        private PlayerData _selectedPlayerData;
        private ProfessionButton _selectedProfessionButton;
        private CharacterEntity _characterEntity;

        #region UI components
        public GameObject core;
        private CommonTopBarView _commonTopBarView;

        public Text changeProfessionText;
        public Button changeProfessionButton;
        public Text professionNameText;
        public Text playerLevelText;
        public Text playerLevelMaxText;
        public Slider playerExpSlider;
        public Text playerExpPercentText;
        public Text professionStrenthText;

        public Text hpTitleText;
        public Text attackTitleText;
        public Text defenceTitleText;
        public Text speedTitleText;
        public Text critTitleText;
        public Text dodgeTitleText;
        public Text blockTitleText;
        public Text hitTitleText;

        public Text hpText;
        public Text attackText;
        public Text defenceText;
        public Text speedText;
        public Text critText;
        public Text dodgeText;
        public Text blockText;
        public Text hitText;

        public Text activeSkillsTitleText;
        public Text passiveSkillsTitleText;

        public Text selectProfessionTitleText;

        public Transform professionButtonsRootTransform;
        public ProfessionButton professionButtonPrefab;
        public GameObject selectedProfessionGlowGameObject;

        private bool _isSkillsPanelShowing = true;
        public GameObject skillsPanelShowHideArrow;
        public GameObject skillsPanelRootGameObject;
        public GameObject strengthenBtnGO;
        public Transform skillPanelShowPosTransform;
        public Transform skillPanelHidePosTransform;
        public Transform roleModelRootTransform;
        public SkillDesButton[] skillDesButton;
        #endregion



        void Start()
        {
            Init();
            BindDelegate();
        }

        void OnDestroy()
        {
            DespawnCharacter();
            UnbindDelegate();
        }

        void Update()
        {
            if (_selectedProfessionButton != null)
            {
                RefreshSelectedProfessionButtonIndicator();
            }
        }

        private void BindDelegate()
        {
            PlayerProxy.instance.onPlayerInfoUpdateDelegate += OnPlayerInfoUpdateHandler;
        }

        private void UnbindDelegate()
        {
            PlayerProxy.instance.onPlayerInfoUpdateDelegate -= OnPlayerInfoUpdateHandler;
        }

        private void Init()
        {
            string title = Localization.Get("ui.change_profession_view.change_profession_title");
            _commonTopBarView = CommonTopBarView.CreateNewAndAttachTo(core.transform);
            _commonTopBarView.SetAsCommonStyle(title, ClickBackHandler, true, true, true, true);

            changeProfessionText.text = Localization.Get("ui.change_profession_view.change_profession");
            professionStrenthText.text = Localization.Get("ui.change_profession_view.profession_strenthen");
            hpTitleText.text = Localization.Get("ui.change_profession_view.health_point");
            attackTitleText.text = Localization.Get("ui.change_profession_view.attack");
            defenceTitleText.text = Localization.Get("ui.change_profession_view.defense");
            speedTitleText.text = Localization.Get("ui.change_profession_view.speed");
            critTitleText.text = Localization.Get("ui.change_profession_view.critial");
            dodgeTitleText.text = Localization.Get("ui.change_profession_view.dodge");
            blockTitleText.text = Localization.Get("ui.change_profession_view.block");
            hitTitleText.text = Localization.Get("ui.change_profession_view.hit");
            activeSkillsTitleText.text = Localization.Get("ui.change_profession_view.active_skills");
            passiveSkillsTitleText.text = Localization.Get("ui.change_profession_view.passive_skills");
            selectProfessionTitleText.text = Localization.Get("ui.change_profession_view.select_profession_title");

            _selectedPlayerData = GameProxy.instance.PlayerInfo.playerData;
            RegenerateProfessionButtons();
            Refresh();
        }

        private void RegenerateProfessionButtons()
        {
            ProfessionButton[] professionButtons = professionButtonsRootTransform.GetComponentsInChildren<ProfessionButton>();
            int professionButtonCount = professionButtons.Length;
            for (int i = 0; i < professionButtonCount; i++)
            {
                if (professionButtons[i] != professionButtonPrefab)
                {
                    DestroyImmediate(professionButtons[i].gameObject);
                }
            }

            professionButtonPrefab.gameObject.SetActive(true);
            ProfessionButton currentPlayerProfessionButton = GameObject.Instantiate<ProfessionButton>(professionButtonPrefab);
            currentPlayerProfessionButton.SetPlayerData(GameProxy.instance.PlayerInfo.playerData);
            currentPlayerProfessionButton.transform.SetParent(professionButtonsRootTransform, false);

            List<PlayerData> changeProfessionPlayerList = PlayerData.GetChangeProfessionPlayerDataList(GameProxy.instance.PlayerInfo.playerData);
            int changeProfessionPlayerListCount = changeProfessionPlayerList.Count;
            for (int i = 0; i < changeProfessionPlayerListCount; i++)
            {
                ProfessionButton professionButton = GameObject.Instantiate<ProfessionButton>(professionButtonPrefab);
                professionButton.SetPlayerData(changeProfessionPlayerList[i]);
                professionButton.transform.SetParent(professionButtonsRootTransform, false);
            }
            professionButtonPrefab.gameObject.SetActive(false);
            professionButtonsRootTransform.localPosition = new Vector3(0, professionButtonsRootTransform.localPosition.y, professionButtonsRootTransform.localPosition.z);

            _selectedProfessionButton = currentPlayerProfessionButton;
        }

        private void ReorderProfessionButtons()
        {
            _selectedProfessionButton.transform.SetAsFirstSibling();
        }

        private void RefreshSelectedProfessionButtonIndicator()
        {
            selectedProfessionGlowGameObject.transform.position = _selectedProfessionButton.transform.position;
            selectedProfessionGlowGameObject.transform.SetAsLastSibling();
        }

        private void RefreshPlayerLevelAndAttributes(PlayerInfo playerInfo)
        {
            int playerLevel = playerInfo.level;
            int playerMaxLevel = GlobalData.GetGlobalData().playerLevelMax;
            float playerExpPercentToNextLevel = PlayerUtil.GetPlayerExpPercentToNextLevel(playerInfo);
            playerLevelText.text = playerLevel.ToString();
            playerLevelMaxText.text = playerMaxLevel.ToString(); ;
            playerExpSlider.value = playerExpPercentToNextLevel;
            playerExpPercentText.text = string.Format(Localization.Get("common.percent"), ((int)(playerExpPercentToNextLevel * 100)));

            Dictionary<RoleAttributeType, RoleAttribute> playerAttributeDictionary = PlayerUtil.CalcPlayerAttributesDic(playerInfo);
            hpText.text = playerAttributeDictionary[RoleAttributeType.HP].ValueString;

            RoleAttackAttributeType roleAttackAttributeType = CharacterUtil.GetRoleAttackAttributeType(playerInfo.heroData.roleType);
            int offence = 0;
            if (roleAttackAttributeType == RoleAttackAttributeType.PhysicalAttack)
            {
                offence = (int)playerAttributeDictionary[RoleAttributeType.NormalAtk].value;
            }
            else if (roleAttackAttributeType == RoleAttackAttributeType.MagicalAttack)
            {
                offence = (int)playerAttributeDictionary[RoleAttributeType.MagicAtk].value;
            }
            attackText.text = offence.ToString();

            defenceText.text = playerAttributeDictionary[RoleAttributeType.Normal_Def].ValueString;
            speedText.text = playerAttributeDictionary[RoleAttributeType.Speed].ValueString;

            float crit = playerAttributeDictionary.ContainsKey(RoleAttributeType.Crit) ? playerAttributeDictionary[RoleAttributeType.Crit].value : 0;
            float dodge = playerAttributeDictionary.ContainsKey(RoleAttributeType.Dodge) ? playerAttributeDictionary[RoleAttributeType.Dodge].value : 0;
            float block = playerAttributeDictionary.ContainsKey(RoleAttributeType.Block) ? playerAttributeDictionary[RoleAttributeType.Block].value : 0;
            float hit = playerAttributeDictionary.ContainsKey(RoleAttributeType.Hit) ? playerAttributeDictionary[RoleAttributeType.Hit].value : 0;
            critText.text = string.Format(Localization.Get("common.percent"), crit);
            dodgeText.text = string.Format(Localization.Get("common.percent"), dodge);
            blockText.text = string.Format(Localization.Get("common.percent"), block);
            hitText.text = string.Format(Localization.Get("common.percent"), hit);
        }

        private void DespawnCharacter()
        {
            if (_characterEntity)
                Pool.Controller.PoolController.instance.Despawn(_characterEntity.name, _characterEntity);
            _characterEntity = null;
        }

        private void RefreshPlayerModel(PlayerInfo playerInfo)
        {
            DespawnCharacter();
            TransformUtil.ClearChildren(roleModelRootTransform, true);
            professionNameText.text = Localization.Get(playerInfo.heroData.name);
            _characterEntity = CharacterEntity.CreatePlayerEntityAsUIElement(playerInfo, roleModelRootTransform, false, true);
        }

        private void Refresh()
        {
            PlayerInfo selectedPlayerInfo = null;
            if (PlayerProxy.instance.IsPlayerUnlocked((int)_selectedPlayerData.Id))
            {
                selectedPlayerInfo = PlayerProxy.instance.GetPlayerDataCorrespondingPlayerInfo((int)_selectedPlayerData.Id);
                changeProfessionText.text = Localization.Get("ui.change_profession_view.change_profession");
                //				strengthenBtnGO.SetActive(true);
            }
            else
            {
                //				strengthenBtnGO.SetActive(false);
                PlayerInfo playerInfo = GameProxy.instance.PlayerInfo;
                uint mockPlayerInstanceID = 0;
                selectedPlayerInfo = new PlayerInfo(mockPlayerInstanceID, _selectedPlayerData.Id, playerInfo.hairCutIndex, playerInfo.hairColorIndex, playerInfo.faceIndex, playerInfo.skinIndex, GameProxy.instance.AccountName);
                changeProfessionText.text = Localization.Get("ui.change_profession_view.transfer_profession");
            }


            bool shouldShowChangeProfessionButton = _selectedPlayerData.Id != GameProxy.instance.PlayerInfo.playerData.Id;
            changeProfessionButton.gameObject.SetActive(shouldShowChangeProfessionButton);

            RefreshPlayerLevelAndAttributes(selectedPlayerInfo);
            RefreshPlayerModel(selectedPlayerInfo);
            RefreshSkill(selectedPlayerInfo);
        }
        private void RefreshSkill(PlayerInfo playerInfo)
        {

            SkillData skillData1 = SkillData.GetSkillDataById(playerInfo.heroData.skillId1);
            if (skillData1 != null)
            {
                skillDesButton[0].gameObject.SetActive(true);
                skillDesButton[0].SetSkillInfo(playerInfo.heroData.skillId1, playerInfo.advanceLevel, (int)playerInfo.heroData.starMin);
				skillDesButton[0].GetComponent<Image>().SetSprite(ResMgr.instance.Load<Sprite>(ResPath.GetSkillIconPath(skillData1.skillIcon)));
            }
            else
            {
                skillDesButton[0].gameObject.SetActive(false);
            }
            SkillData skillData2 = SkillData.GetSkillDataById(playerInfo.heroData.skillId2);
            if (skillData2 != null)
            {
                skillDesButton[1].gameObject.SetActive(true);
                skillDesButton[1].SetSkillInfo(playerInfo.heroData.skillId2, playerInfo.advanceLevel, (int)playerInfo.heroData.starMin);
				skillDesButton[1].GetComponent<Image>().SetSprite(ResMgr.instance.Load<Sprite>(ResPath.GetSkillIconPath(skillData2.skillIcon)));
            }
            else
            {
                skillDesButton[1].gameObject.SetActive(false);
            }
            SkillData passiveSkillData = SkillData.GetSkillDataById(playerInfo.heroData.passiveId1);
            if (passiveSkillData != null)
            {
                skillDesButton[2].gameObject.SetActive(true);
                skillDesButton[2].SetSkillInfo(playerInfo.heroData.passiveId1, playerInfo.advanceLevel, (int)playerInfo.heroData.starMin);
				skillDesButton[2].GetComponent<Image>().SetSprite(ResMgr.instance.Load<Sprite>(ResPath.GetSkillIconPath(passiveSkillData.skillIcon)));
            }
            else
            {
                skillDesButton[2].gameObject.SetActive(false);
            }
        }
        private void OnPlayerInfoUpdateHandler()
        {
            RefreshPlayerModel(GameProxy.instance.PlayerInfo);
            RegenerateProfessionButtons();
            Refresh();
        }

        #region UI event handlers
        public void ClickBackHandler()
        {
            UIMgr.instance.Close(PREFAB_PATH);
        }

        public void ClickShowHideSkillsPanel()
        {
            if (_isSkillsPanelShowing)
            {
                LeanTween.moveLocalY(skillsPanelRootGameObject, skillPanelHidePosTransform.localPosition.y, 0.2f);
                skillsPanelShowHideArrow.transform.localRotation = Quaternion.Euler(0, 0, 180);
            }
            else
            {
                LeanTween.moveLocalY(skillsPanelRootGameObject, skillPanelShowPosTransform.localPosition.y, 0.2f);
                skillsPanelShowHideArrow.transform.localRotation = Quaternion.Euler(0, 0, 0);
            }
            _isSkillsPanelShowing = !_isSkillsPanelShowing;
        }

        public void ClickProfessionButton(ProfessionButton professionButton)
        {
            _selectedPlayerData = professionButton.PlayerData;
            _selectedProfessionButton = professionButton;
            RefreshSelectedProfessionButtonIndicator();
            Refresh();
            //			if(!PlayerProxy.instance.IsPlayerUnlocked((int)_selectedPlayerData.Id))
            //			{
            //				ActivateProfessionView activateProfessionView = UIMgr.instance.Open<ActivateProfessionView>(ActivateProfessionView.PREFAB_PATH);
            //				activateProfessionView.SetPlayerData(_selectedPlayerData);
            //			}
        }

        public void ClickChangeProfessionButtonHandler()
        {
            if (PlayerProxy.instance.IsPlayerUnlocked((int)_selectedPlayerData.Id))
            {
                PlayerInfo playerInfo = PlayerProxy.instance.GetPlayerDataCorrespondingPlayerInfo((int)_selectedPlayerData.Id);
                PlayerController.instance.CLIENT2LOBBY_PLAYER_CHANGE_REQ((int)playerInfo.instanceID);
            }
            else
            {
                ActivateProfessionView.Open(_selectedPlayerData);
            }
        }
        public void ClickChangeProfessionStrengthenButtonHandler()
        {
            if (PlayerProxy.instance.IsPlayerUnlocked((int)_selectedPlayerData.Id))
            {
                PlayerInfo selectedPlayerInfo = PlayerProxy.instance.GetPlayerDataCorrespondingPlayerInfo((int)_selectedPlayerData.Id);
                Logic.UI.HeroStrengthen.View.HeroStrengthenView.Open(selectedPlayerInfo);
            }

        }
        #endregion
    }
}